### PR TITLE
Alc 313 min amount

### PR DESCRIPTION
**Pull request by Dire-0x**
- add method on the Trade class to calculate min trade amounts `estimateMinTradeAmounts`
- Fix price impact calculations

### CODE REVIEW

File: src/entities/trade.ts [hunk 1]
```diff
@@ -576,4 +578,96 @@ export class Trade {
...
+      }
+      if (pair.reserve0.equalTo(ZERO) || pair.reserve1.equalTo(ZERO)) continue
+
+      
+      try {
+        if (etherIn){
+          minTokenAmountIn = CurrencyAmount.ether(exactInBribe)
+          ;[minTokenAmountOut] = pair.getInputAmount(amountOut0, Exchange.UNI)
+        } else if (etherOut) {
+          minTokenAmountOut = CurrencyAmount.ether(exactInBribe)
```
Comment (swgz9) [inline src/entities/trade.ts]: Should this be `exactOutBribe`?

File: src/entities/trade.ts [hunk 2]
```diff
@@ -576,4 +578,96 @@ export class Trade {
...
+    currencyIn: Currency,
+    currencyOut: Currency, 
+    gasPriceToBeat: BigintIsh,
+    minerBribeMargin: BigintIsh,
+    { maxHops = 3 }: BestTradeOptions = {}
+  ): BribeEstimation | undefined {
+    const etherIn = currencyIn === ETHER
+    const etherOut = currencyOut === ETHER
+
+    const exactInGas = estimatedGasForMethod(Trade.methodNameForTradeType(TradeType.EXACT_INPUT, etherIn, etherOut), maxHops.toString())j
```
Comment (swgz9) [inline src/entities/trade.ts]: `estimatedGasForMethod` in the SDK still has `const factor = parseBigintIsh('0') // TODO: change this` I guess we should re-visit that factor or keep it in mind

**Review by swgz9 — APPROVED** [review]

File: src/entities/trade.ts [hunk 3]
```diff
@@ -576,4 +579,77 @@ export class Trade {
...
+    for (let i = 0; i < pairs.length; i++){
+      const pair = pairs[i]
+      // pair irrelevant
+      if (!pair.token0.equals(amountOut.token) && !pair.token1.equals(amountOut.token)) continue 
+      if (pair.reserve0.equalTo(ZERO) || pair.reserve1.equalTo(ZERO)) continue
+
+      try {
+        if (etherIn){
+          minTokenAmountIn = CurrencyAmount.ether(exactInBribe)
+          ;[minTokenAmountOut] = pair.getInputAmount(amountOut, Exchange.UNI)
```
Comment (swgz9) [inline src/entities/trade.ts]: Just saw that you are using `Exchange.UNI` can we make `exchange` a parameter and call accordingly where we need this method?

File: src/entities/trade.ts (hunk 2 above)
Comment (Dire-0x) [reply src/entities/trade.ts]: We still need to figure out what that factor is, it will be a constant. Ill make a task to do this before MVP

**Pull request merged**